src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs

```diff
@@ -107,7 +129,24 @@ public OtlpExporterOptions()
         /// <summary>
         /// Gets or sets the the OTLP transport protocol. Supported values: Grpc and HttpProtobuf.
         /// </summary>
-        public OtlpExportProtocol Protocol { get; set; } = OtlpExportProtocol.Grpc;
+        public OtlpExportProtocol Protocol
+        {
+            get => this.protocol;
+            set
+            {
+                this.protocol = value;
+
+                switch (value)
```

Comment:
Should this switch be wrapped in a check to `if (!this.ProgrammaticallyModifiedEndpoint)`?

Reply:
`this.SetEndpoint` has:

```
                if (!this.isCustomEndpointSet)
                {
                    this.endpoint = value;
                }
```

Reply:
@CodeBlanch Would you prefer to refactor this code to:

```csharp
          if (!this.isCustomEndpointSet)
          {
                switch (value)                
                {
                    case OtlpExportProtocol.Grpc:
                        this.SetEndpoint(new Uri(DefaultGrpcEndpoint));
                        break;
                    case OtlpExportProtocol.HttpProtobuf:
                        this.SetEndpoint(new Uri(DefaultHttpProtobufEndpoint));
                        break;
                }                {
                    case OtlpExportProtocol.Grpc:
                        this.SetEndpoint(new Uri(DefaultGrpcEndpoint));
                        break;
                    case OtlpExportProtocol.HttpProtobuf:
                        this.SetEndpoint(new Uri(DefaultHttpProtobufEndpoint));
                        break;
                }
            }
```

Get rid of `bool custom` parameter in the `SetEndpoint` and maybe also rename the method to `SetCustomEndpoint` ?

Reply:
I like that might make it a bit easier to follow what is going on. Would it help if we just left Endpoint `null` unless the user sets it explicitly? Then if endpoint is `null` on exporter startup then we could construct one from the default(s).

Reply:
@CodeBlanch, I have considered to remove SetEndopint method and extend Protocol setter. Then we have two places modifying the same field. IMO it will be even harder to follow.

If we consider the null-value and extend startup method - we need to do it in 2-3 places (traces/metric/logs) separately. I am not strongly against this, but the logic leeks outside the settings class.

Reply:
> If we consider the null-value and extend startup method - we need to do it in 2-3 places (traces/metric/logs) separately. I am not strongly against this, but the logic leeks outside the settings class.

We could probably have a method on `OtlpExporterOption`s to encapsulate all that logic (endpoint resolution) or use the `AppendExportPath` extension?

I don't have any issue with the way it is currently being done. Happy to approve as-is.

Just some oddness standing out for me I thought I would raise.

* Protocol property setter is mutating the Endpoint property. That is a code smell IMO. I think calling a method, user expects class state to potentially mutate. But property set affecting other properties seems unusual? I feel like setting the values and choosing the final endpoint are two different units.

* Mutating the options from environment variables seems to be causing some pain. Could the whole thing be simplified if we had 2 options classes? 1 from the user. 1 constructed from environment variables. Some logic to build endpoint considering both? That logic would be something like: If user specified something on options, respect it. If not, check options created from environment variables. Otherwise fallback to defaults.

Reply:
Along the thought of endpoint just being null to begin with. Could it look like this:

```csharp
        private Uri endpoint = null;
        public Uri Endpoint
        {
            get
            {
                if (endpoint == null) { ... return the default based on protocol ... }
                else return endpoint;
            }
            set
            {
                this.endpoint = value;
            }
        }
```

Reply:
@alanwest, @CodeBlanch thanks for the input. Last version with null starting commited.